README.md
distortosExamples [![Build Status](https://travis-ci.org/DISTORTEC/distortosExamples.svg?branch=master)](https://travis-ci.org/DISTORTEC/distortosExamples)
==========================

Examples for [distortos](http://distortos.org/) - object-oriented C++ RTOS for microcontrollers

Configuration & building
------------------------

To configure & build *distortosExamples* you need:
- [CMake](https://cmake.org/) (version 3.7 or later);
- [a build tool supported by CMake](https://cmake.org/cmake/help/latest/manual/cmake-generators.7.html#manual:cmake-generators(7)) -
it is highly recommended to use [Ninja](https://ninja-build.org/);
- [arm-none-eabi bleeding-edge-toolchain](https://github.com/FreddieChopin/bleeding-edge-toolchain) (*GCC* version 5 or
later);

Build infrastructure of *distortos* tries to follow typical *CMake* cross-compiling workflow, which means that you
always have to use a so-called *toolchain file*. Toolchain files in *distortos* also serve another purpose - they select
the board which is going to be used by your application.

1. Download source package of *distortosExamples* in
[zip](https://github.com/DISTORTEC/distortosExamples/archive/master.zip) or
[tar.gz](https://github.com/DISTORTEC/distortosExamples/archive/master.tar.gz) format and extract it;
2. Download source package of *distortos* in [zip](https://github.com/DISTORTEC/distortos/archive/master.zip) or
[tar.gz](https://github.com/DISTORTEC/distortos/archive/master.tar.gz) format and extract it inside
*distortosExamples* extracted in step 1, for example to `distortos/`;
3. Adjust relative path to *distortos* in `add_subdirectory(<path-to-distortos>)` from top-level `CMakeLists.txt`;
4. Create a build folder, for example `output`;
5. From within the build folder, initialize it with *CMake*, for example with
`cmake .. -DCMAKE_TOOLCHAIN_FILE=../distortos/source/board/ST_STM32F4DISCOVERY/Toolchain-ST_STM32F4DISCOVERY.cmake -GNinja`
if you want a default configuration or
`cmake -C../configurations/ST_STM32F4DISCOVERY/distortosConfiguration.cmake .. -GNinja` if you want to start from a
saved configuration;
6. Edit *distortos* configuration with a tool of your choice, for example `cmake-gui ..` (a GUI application) or
`ccmake ..` (*curses*-based application);
7. Execute selected build tool, for example `ninja` or `ninja -v` if you want to see all command lines while building;

You can obviously replace steps 1-3 with
`git clone --recursive https://github.com/DISTORTEC/distortosExamples`.

Steps 4-6 can be all done from within `cmake-gui`. After starting the application use *Browse Source...* button to
select the folder with *distortosExamples* and *Browse Build...* button to select the build folder. Then click
on *Configure* button. In the *CMakeSetup* window which appears select the generator of your choice and make sure that
*Specify toolchain file for cross-compiling* is selected before going any further. Click *Next* and specify the
toolchain file (which also selects the board), for example
`<source-folder>/distortos/source/board/ST_STM32F4DISCOVERY/Toolchain-ST_STM32F4DISCOVERY.cmake` and click *Finish*
button.

### tl;dr

    $ wget https://github.com/DISTORTEC/distortosExamples/archive/master.tar.gz -O distortosExamples-master.tar.gz
    $ tar -xf distortosExamples-master.tar.gz
    $ wget https://github.com/DISTORTEC/distortos/archive/master.tar.gz -O distortos-master.tar.gz
    $ tar -xf distortos-master.tar.gz
    $ mv -T distortos-master/ distortosExamples-master/distortos/
    $ cd distortosExamples-master
    $ mkdir output
    $ cd output
    $ cmake .. -DCMAKE_TOOLCHAIN_FILE=../distortos/source/board/ST_STM32F4DISCOVERY/Toolchain-ST_STM32F4DISCOVERY.cmake -GNinja
    $ cmake-gui ..
    $ ninja

or

    $ wget https://github.com/DISTORTEC/distortosExamples/archive/master.tar.gz -O distortosExamples-master.tar.gz
    $ tar -xf distortosExamples-master.tar.gz
    $ wget https://github.com/DISTORTEC/distortos/archive/master.tar.gz -O distortos-master.tar.gz
    $ tar -xf distortos-master.tar.gz
    $ mv -T distortos-master/ distortosExamples-master/distortos/
    $ cd distortosExamples-master
    $ mkdir output
    $ cd output
    $ cmake -C../configurations/ST_STM32F4DISCOVERY/distortosConfiguration.cmake .. -GNinja
    $ cmake-gui ..
    $ ninja


CHANGELOG.md
Change Log
==========

All notable changes to this project will be documented in this file.

[Unreleased](https://github.com/DISTORTEC/distortosExamples/compare/v20190505...HEAD)
-------------------------------------------------------------------------------------

### Added

- Configuration for [NUCLEO-F767ZI](https://www.st.com/en/evaluation-tools/nucleo-f767zi.html) board with *STM32F7*
chip.

[20190505](https://github.com/DISTORTEC/distortosExamples/compare/v20180701...v20190505) - 2019-05-05
-----------------------------------------------------------------------------------------------------

### Added

- Configurations for [32F072BDISCOVERY](https://www.st.com/en/evaluation-tools/32f072bdiscovery.html) board with
*STM32F0* chip.

### Changed

- Implemented full support for configuring and building with *CMake*. The new *CMake* workflow does not need *Kconfig*
or any shell tools - just *CMake* (version 3.7 or later), build tool (it is recommended to use *Ninja*) and
*arm-none-eabi bleeding-edge-toolchain* (*GCC* version 5 or later). Check `README.md` for more details about usage.
- Replace all uses of `CONFIG_BOARD_LEDS_ENABLE` macro with `DISTORTOS_BOARD_LEDS_ENABLE`.

### Removed

- Removed support for configuring with *Kconfig* and building with *make*. Both of these tasks are now handled by
*CMake*.

[20180701](https://github.com/DISTORTEC/distortosExamples/compare/v20170914...v20180701) - 2018-07-01
-----------------------------------------------------------------------------------------------------

### Added

- Configurations for [NUCLEO-L432KC](http://www.st.com/en/evaluation-tools/nucleo-l432kc.html) and
[NUCLEO-L476RG](http://www.st.com/en/evaluation-tools/nucleo-l476rg.html) boards with *STM32L4* chips.
- Configuration for [NUCLEO-F446RE](http://www.st.com/en/evaluation-tools/nucleo-f446re.html) board with *STM32F4* chip.
- *CMake*-based build system. At this moment all configuration is still done with *Kconfig* - *CMake* loads selected
`distortosConfiguration.mk` and only deals with compilation. Typical use case involves following steps: select
configuration with `make configure CONFIG_PATH=...`, create output folder of your choice (`mkdir output`) and enter it
(`cd output`), configure compilation with
`cmake .. -DCMAKE_TOOLCHAIN_FILE=../distortos/cmake/Toolchain-arm-none-eabi.cmake` and finally start the build with
`make`. You can obviously use other
[*CMake* generators](https://cmake.org/cmake/help/latest/manual/cmake-generators.7.html), e.g. *Ninja*, *Eclipse CDT4*
project, ... *CMake*-based build system will only support configurations with proper board, i.e. "Custom board" choice
for "Board" in *Kconfig* is not supported.

### Changed

- Replace all uses of `distortos::board::totalLeds` constant and `DISTORTOS_BOARD_TOTAL_LEDS` macro with
`distortos::board::ledsCount` and `DISTORTOS_BOARD_LEDS_COUNT`.

### Removed

- Removed *tup* build infrastructure - *tup* is no longer supported for building *distortos*.

[20170914](https://github.com/DISTORTEC/distortosExamples/compare/v20170311...v20170914) - 2017-09-14
-----------------------------------------------------------------------------------------------------

### Added

- Configuration for [NUCLEO-L073RZ](http://www.st.com/en/evaluation-tools/nucleo-l073rz.html) board with *STM32L0* chip.
- Configuration for [32F769IDISCOVERY](http://www.st.com/en/evaluation-tools/32f769idiscovery.html) board with *STM32F7*
chip.

[20170311](https://github.com/DISTORTEC/distortosExamples/compare/v20161124...v20170311) - 2017-03-11
-----------------------------------------------------------------------------------------------------

### Added

- Configuration for [32F746GDISCOVERY](http://www.st.com/en/evaluation-tools/32f746gdiscovery.html) board with *STM32F7*
chip.

### Changed

- Enabled function context checking and all stack overflow checks in configuration files.
- Changed number of queued signals and signal actions for `main()` thread to 8 in all `distortosConfiguration.mk` files.
- Reduced size of stack for interrupts to 1kB in all configurations.
- Replace all uses of `CONFIG_BOARD_TOTAL_LEDS` macro with `DISTORTOS_BOARD_TOTAL_LEDS`.

[20161124](https://github.com/DISTORTEC/distortosExamples/compare/v20160503...v20161124) - 2016-11-24
-----------------------------------------------------------------------------------------------------

This is just a snapshot with version *0.3.0* of *distortos*.

[20160503](https://github.com/DISTORTEC/distortosExamples/compare/v20160226...v20160503) - 2016-05-03
-----------------------------------------------------------------------------------------------------

### Added

- Configurations for [NUCLEO-F091RC](http://www.st.com/web/catalog/tools/PF260944) board with *STM32F0* chip,
[NUCLEO-F103RB](http://www.st.com/nucleoF103RB-pr) board with *STM32F1* chip and
[NUCLEO-F429ZI](http://www.st.com/web/catalog/tools/PF262637) board with *STM32F4* chip.
- Ability to selectively enable/disable examples via *Kconfig* menus.
- *staticSoftwareTimerBlinker* example application, which demonstrates the most basic aspects of static software timers:
creating and starting them, passing arguments to timers's function by reference and by value.

### Changed

- Change "Interrupt priority disabled in critical sections" (*ARCHITECTURE_ARMV7_M_KERNEL_BASEPRI*) option to 0 in all
configurations, so they use *ARMv7-M*'s *PRIMASK* to disable all interrupts during critical sections. This is the new
default setting in *distortos*, as this is a more conservative (and thus safer for a beginner) approach, which doesn't
require any special configuration of interrupt priorities (default value is 0).

20160226 - 2016-02-26
---------------------

### Added

- Build infrastructure from [distortosTemplateSubfolder](https://github.com/DISTORTEC/distortosTemplateSubfolder).
- *staticThreadBlinker* example application, which demonstrates the most basic aspects of static threads: creating and

Tuprules.lua
--
-- file: Tuprules.lua
--
-- author: Copyright (C) 2016 Kamil Szczygiel http://www.distortec.com http://www.freddiechopin.info
--
-- This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not
-- distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
--

------------------------------------------------------------------------------------------------------------------------
-- path to distortos subfolder
------------------------------------------------------------------------------------------------------------------------

local DISTORTOS_PATH = "distortos"

------------------------------------------------------------------------------------------------------------------------
-- reuse Tuprules.lua from distortos subfolder
------------------------------------------------------------------------------------------------------------------------

TOP = tup.getcwd() .. "/"
TOP_NODE = tup.nodevariable(".")
tup.include(DISTORTOS_PATH .. "/Tuprules.lua")

------------------------------------------------------------------------------------------------------------------------
-- start group of <objects> for template application
------------------------------------------------------------------------------------------------------------------------

startObjectsGroup()


threadExamples/staticThreadBlinker/Tupfile.lua
--
-- file: Tupfile.lua
--
-- author: Copyright (C) 2014-2016 Kamil Szczygiel http://www.distortec.com http://www.freddiechopin.info
--
-- This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not
-- distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
--

if CONFIG_STATICTHREADBLINKER_ENABLE == "y" then

	CXXFLAGS += STANDARD_INCLUDES
	CXXFLAGS += CHIP_INCLUDES
	CXXFLAGS += BOARD_INCLUDES
	
	tup.include(DISTORTOS_TOP .. "compile.lua")
	
	local filename = OUTPUT .. tup.getrelativedir(TOP) .. "/staticThreadBlinker"
	local elfFilename = filename .. ".elf"
	local hexFilename = filename .. ".hex"
	local binFilename = filename .. ".bin"
	local dmpFilename = filename .. ".dmp"
	local lssFilename = filename .. ".lss"
	
	link(elfFilename, OUTPUT .. "libdistortos.a", "<objects>", LDSCRIPT)
	size(elfFilename)
	hex(elfFilename, hexFilename)
	bin(elfFilename, binFilename)
	dmp(elfFilename, dmpFilename)
	lss(elfFilename, lssFilename)

end	-- if CONFIG_STATICTHREADBLINKER_ENABLE == "y" then


threadExamples/dynamicThreadBlinker/dynamicThreadBlinker.cpp
/**
 * \file
 * \brief dynamicThreadBlinker example application
 *
 * \author Copyright (C) 2016-2019 Kamil Szczygiel http://www.distortec.com http://www.freddiechopin.info
 *
 * \par License
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not
 * distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

#include "distortos/distortosConfiguration.h"

#if defined(DISTORTOS_BOARD_LEDS_ENABLE)

#include "distortos/board/leds.hpp"

#include "distortos/chip/ChipOutputPin.hpp"

#endif	// defined(DISTORTOS_BOARD_LEDS_ENABLE)

#include "distortos/DynamicThread.hpp"
#include "distortos/ThisThread.hpp"

namespace
{

/*---------------------------------------------------------------------------------------------------------------------+
| local functions
+---------------------------------------------------------------------------------------------------------------------*/

#if defined(DISTORTOS_BOARD_LEDS_ENABLE)

/**
 * \brief LED blinking function
 *
 * Constantly toggles state of \a led and waits for half of \a periodMs.
 *
 * \param [in] led is a reference to distortos::devices::OutputPin object which will be toggled by this function
 * \param [in] periodMs is a full (on -> off -> on) period of toggling, milliseconds
 */

void ledBlinkerFunction(distortos::devices::OutputPin& led, const std::chrono::milliseconds periodMs)
{
	while (1)
	{
		led.set(!led.get());	// invert state of LED
		distortos::ThisThread::sleepFor(periodMs / 2);
	}
}

#endif	// defined(DISTORTOS_BOARD_LEDS_ENABLE)

/**
 * \brief Boolean variable "blinking" function
 *
 * Constantly toggles value of \a variable and waits for half of \a periodMs. This function is meant as a demonstration
 * for configuration which either has no LEDs or doesn't enable them.
 *
 * \param [in] variable is a reference to bool variable which will be toggled by this function
 * \param [in] periodMs is a full (true -> false -> true) period of toggling, milliseconds
 */

void variableBlinkerFunction(bool& variable, const std::chrono::milliseconds periodMs)
{
	while (1)
	{
		variable = !variable;	// invert state of variable
		distortos::ThisThread::sleepFor(periodMs / 2);
	}
}

}	// namespace

/*---------------------------------------------------------------------------------------------------------------------+
| global functions
+---------------------------------------------------------------------------------------------------------------------*/

/**
 * \brief Main code block of dynamicThreadBlinker application
 *
 * This example application tries to demonstrate the most basic (and one more advanced) aspects of dynamic threads:
 * - creating and starting them,
 * - passing arguments to thread's function by reference and by value,
 * - detaching them.
 *
 * If the configured board has no LEDs or if they are not enabled in the configuration, only one dummy thread is
 * created - it "blinks" a boolean variable instead of a real LED. Otherwise up to four additional threads are created
 * (but no more than the number of LEDs on the board - distortos::board::ledsCount) - each one blinks its own LED (which
 * was passed to the thread's function by reference) with provided period (passed by value). The periods of blinking are
 * slightly different for each thread, so if there are multiple LEDs they are not in sync with each other. The periods
 * are actually prime numbers, so they create very long "global" period (in which the whole pattern repeats).
 *
 * As all created threads never terminate (their functions have infinite loops and never return) and main thread has no
 * interest in synchronizing with them, they are immediately detached right after being constructed and started. Main
 * thread doesn't have anything to do after that, so it just sleeps for long periods in a loop.
 *
 * \return doesn't return
 */

int main()
{
	bool variable {};
	// create, immediately start and immediately detach dynamic thread with 1024 bytes of stack, low priority (1),
	// variableBlinkerFunction() will get variable by reference and period by value
	distortos::makeAndStartDynamicThread({1024, 1}, variableBlinkerFunction, std::ref(variable),
			std::chrono::milliseconds{401}).detach();

#if defined(DISTORTOS_BOARD_LEDS_ENABLE)

	static const std::array<std::chrono::milliseconds, 4> periods
	{
			std::chrono::milliseconds{397},
			std::chrono::milliseconds{389},
			std::chrono::milliseconds{383},
			std::chrono::milliseconds{379},
	};

	constexpr size_t threadsCount {periods.size() < distortos::board::ledsCount ?
			periods.size() : distortos::board::ledsCount};
	// create, immediately start and immediately detach dynamic thread with 1024 bytes of stack, low priority (1),
	// ledBlinkerFunction() will get its own LED by reference and period by value
	for (size_t i {}; i < threadsCount; ++i)
		distortos::makeAndStartDynamicThread({1024, 1}, ledBlinkerFunction, std::ref(distortos::board::leds[i]),
				periods[i]).detach();

#endif	// defined(DISTORTOS_BOARD_LEDS_ENABLE)

	while (1)
		distortos::ThisThread::sleepFor(std::chrono::hours{1});

	return 0;
}


softwareTimerExamples/staticSoftwareTimerBlinker/staticSoftwareTimerBlinker.cpp
/**
 * \file
 * \brief staticSoftwareTimerBlinker example application
 *
 * \author Copyright (C) 2016-2017 Kamil Szczygiel http://www.distortec.com http://www.freddiechopin.info
 *
 * \par License
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not
 * distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

#include "distortos/board/leds.hpp"

#include "distortos/chip/ChipOutputPin.hpp"

#include "distortos/ThisThread.hpp"
#include "distortos/StaticSoftwareTimer.hpp"

/*---------------------------------------------------------------------------------------------------------------------+
| global functions
+---------------------------------------------------------------------------------------------------------------------*/

/**
 * \brief Main code block of staticSoftwareTimerBlinker application
 *
 * This example application tries to demonstrate the most basic aspects of static software timers:
 * - creating and starting them,
 * - passing arguments to timers's function by reference and by value.
 *
 * Up to four software timers are created (but no more than the number of LEDs on the board -
 * DISTORTOS_BOARD_LEDS_COUNT) - each configured to switch its own LED (which was passed to the timer's function by
 * reference) off (the "off" state is passed by value). Then - in an endless loop - main thread turns all LEDs on,
 * starts all created timers (with different durations) and sleeps for 1 second.
 *
 * \return doesn't return
 */

int main()
{
#if DISTORTOS_BOARD_LEDS_COUNT >= 1

	auto ledOffTimer0 = distortos::makeStaticSoftwareTimer(&distortos::devices::OutputPin::set,
			std::ref(distortos::board::leds[0]), false);

#endif	// DISTORTOS_BOARD_LEDS_COUNT >= 1
#if DISTORTOS_BOARD_LEDS_COUNT >= 2

	auto ledOffTimer1 = distortos::makeStaticSoftwareTimer(&distortos::devices::OutputPin::set,
			std::ref(distortos::board::leds[1]), false);

#endif	// DISTORTOS_BOARD_LEDS_COUNT >= 2
#if DISTORTOS_BOARD_LEDS_COUNT >= 3

	auto ledOffTimer2 = distortos::makeStaticSoftwareTimer(&distortos::devices::OutputPin::set,
			std::ref(distortos::board::leds[2]), false);

#endif	// DISTORTOS_BOARD_LEDS_COUNT >= 3
#if DISTORTOS_BOARD_LEDS_COUNT >= 4

	auto ledOffTimer3 = distortos::makeStaticSoftwareTimer(&distortos::devices::OutputPin::set,
			std::ref(distortos::board::leds[3]), false);

#endif	// DISTORTOS_BOARD_LEDS_COUNT >= 4

	while (1)
	{
		for (auto& led : distortos::board::leds)
			led.set(true);

#if DISTORTOS_BOARD_LEDS_COUNT >= 1

		ledOffTimer0.start(std::chrono::milliseconds{500});

#endif	// DISTORTOS_BOARD_LEDS_COUNT >= 1
#if DISTORTOS_BOARD_LEDS_COUNT >= 2

		ledOffTimer1.start(std::chrono::milliseconds{550});

#endif	// DISTORTOS_BOARD_LEDS_COUNT >= 2
#if DISTORTOS_BOARD_LEDS_COUNT >= 3

		ledOffTimer2.start(std::chrono::milliseconds{600});

#endif	// DISTORTOS_BOARD_LEDS_COUNT >= 3
#if DISTORTOS_BOARD_LEDS_COUNT >= 4

		ledOffTimer3.start(std::chrono::milliseconds{650});

#endif	// DISTORTOS_BOARD_LEDS_COUNT >= 4

		distortos::ThisThread::sleepFor(std::chrono::seconds{1});
	}

	return 0;
}


threadExamples/staticThreadBlinker/staticThreadBlinker.cpp
/**
 * \file
 * \brief staticThreadBlinker example application
 *
 * \author Copyright (C) 2016-2019 Kamil Szczygiel http://www.distortec.com http://www.freddiechopin.info
 *
 * \par License
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not
 * distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

#include "distortos/distortosConfiguration.h"

#if defined(DISTORTOS_BOARD_LEDS_ENABLE)

#include "distortos/board/leds.hpp"

#include "distortos/chip/ChipOutputPin.hpp"

#endif	// defined(DISTORTOS_BOARD_LEDS_ENABLE)

#include "distortos/StaticThread.hpp"
#include "distortos/ThisThread.hpp"

namespace
{

/*---------------------------------------------------------------------------------------------------------------------+
| local functions
+---------------------------------------------------------------------------------------------------------------------*/

#if defined(DISTORTOS_BOARD_LEDS_ENABLE)

/**
 * \brief LED blinking function
 *
 * Constantly toggles state of \a led and waits for half of \a periodMs.
 *
 * \param [in] led is a reference to distortos::devices::OutputPin object which will be toggled by this function
 * \param [in] periodMs is a full (on -> off -> on) period of toggling, milliseconds
 */

void ledBlinkerFunction(distortos::devices::OutputPin& led, const std::chrono::milliseconds periodMs)
{
	while (1)
	{
		led.set(!led.get());	// invert state of LED
		distortos::ThisThread::sleepFor(periodMs / 2);
	}
}

#endif	// defined(DISTORTOS_BOARD_LEDS_ENABLE)

/**
 * \brief Boolean variable "blinking" function
 *
 * Constantly toggles value of \a variable and waits for half of \a periodMs. This function is meant as a demonstration
 * for configuration which either has no LEDs or doesn't enable them.
 *
 * \param [in] variable is a reference to bool variable which will be toggled by this function
 * \param [in] periodMs is a full (true -> false -> true) period of toggling, milliseconds
 */

void variableBlinkerFunction(bool& variable, const std::chrono::milliseconds periodMs)
{
	while (1)
	{
		variable = !variable;	// invert state of variable
		distortos::ThisThread::sleepFor(periodMs / 2);
	}
}

}	// namespace

/*---------------------------------------------------------------------------------------------------------------------+
| global functions
+---------------------------------------------------------------------------------------------------------------------*/

/**
 * \brief Main code block of staticThreadBlinker application
 *
 * This example application tries to demonstrate the most basic aspects of static threads:
 * - creating and starting them,
 * - passing arguments to thread's function by reference and by value,
 * - joining them.
 *
 * If the configured board has no LEDs or if they are not enabled in the configuration, only one dummy thread is
 * created - it "blinks" a boolean variable instead of a real LED. Otherwise up to four additional threads are created
 * (but no more than the number of LEDs on the board - DISTORTOS_BOARD_LEDS_COUNT) - each one blinks its own LED (which
 * was passed to the thread's function by reference) with provided period (passed by value). The periods of blinking are
 * slightly different for each thread, so if there are multiple LEDs they are not in sync with each other. The periods
 * are actually prime numbers, so they create very long "global" period (in which the whole pattern repeats).
 *
 * Even though all created threads never terminate (their functions have infinite loops and never return), main thread
 * calls Thread::join() for each of them anyway. This is a good and safe practice, which also happens to be the easiest
 * way to suspend main thread in this example application.
 *
 * \return doesn't return
 */

int main()
{
	bool variable {};
	// create and immediately start static thread with 1024 bytes of stack, low priority (1), variableBlinkerFunction()
	// will get variable by reference and period by value
	auto variableBlinkerThread = distortos::makeAndStartStaticThread<1024>(1, variableBlinkerFunction,
			std::ref(variable), std::chrono::milliseconds{401});

#if defined(DISTORTOS_BOARD_LEDS_ENABLE) && DISTORTOS_BOARD_LEDS_COUNT >= 1

	// create and immediately start static thread with 1024 bytes of stack, low priority (1), ledBlinkerFunction() will
	// get its own LED by reference and period by value
	auto ledBlinkerThread0 = distortos::makeAndStartStaticThread<1024>(1, ledBlinkerFunction,
			std::ref(distortos::board::leds[0]), std::chrono::milliseconds{397});

#	if DISTORTOS_BOARD_LEDS_COUNT >= 2

	auto ledBlinkerThread1 = distortos::makeAndStartStaticThread<1024>(1, ledBlinkerFunction,
			std::ref(distortos::board::leds[1]), std::chrono::milliseconds{389});

#		if DISTORTOS_BOARD_LEDS_COUNT >= 3

	auto ledBlinkerThread2 = distortos::makeAndStartStaticThread<1024>(1, ledBlinkerFunction,
			std::ref(distortos::board::leds[2]), std::chrono::milliseconds{383});

#			if DISTORTOS_BOARD_LEDS_COUNT >= 4

	auto ledBlinkerThread3 = distortos::makeAndStartStaticThread<1024>(1, ledBlinkerFunction,
			std::ref(distortos::board::leds[3]), std::chrono::milliseconds{379});

	ledBlinkerThread3.join();

#			endif	// DISTORTOS_BOARD_LEDS_COUNT >= 4

	ledBlinkerThread2.join();

#		endif	// DISTORTOS_BOARD_LEDS_COUNT >= 3

	ledBlinkerThread1.join();

#	endif	// DISTORTOS_BOARD_LEDS_COUNT >= 2

	ledBlinkerThread0.join();

#endif	// defined(DISTORTOS_BOARD_LEDS_ENABLE) && DISTORTOS_BOARD_LEDS_COUNT >= 1

	variableBlinkerThread.join();

	return 0;
}
